### ISSUE TITLE
False positive undefined in function generic bounds

### ISSUE THREAD

**Issue by Fuyukai** [microsoft/pyright#7678]
**Describe the bug**
Function generic bounds with types that are defined later have a false positive undefined error.

**Code or Screenshots**

```python
def fn[T: Unk]() -> None:
    ...

class Unk:
    ...
```

Note that both of the following work fine:

```
type _Unk = Unk

def fn[T: _Unk]() -> None:
    ...

# or
def fn[T: "Unk"]() -> None:
	...
```

Runtime introspection works fine, too:

```
>>> fn.__type_params__[0].__bound__
<class '__main__.Unk'>
```

**VS Code extension or command-line**
CLI, Pyright 1.1.358.

**Comment by erictraut** [comment]
Pyright is working in accordance with [PEP 695](https://peps.python.org/pep-0695/#upper-bound-specification) — at least the way that I had intended when writing the PEP. You can see in the linked section of the PEP that forward references are supported if they are quoted (or deferred evaluation is enabled in Python 3.12 via `from __future__ import annotations`).

Note that the type evaluation technique for an expression differs when evaluation is deferred. When evaluation is not deferred, the code flow graph is used to evaluate the expression _at that point in the code flow_. The difference becomes apparent when a symbol is redefined.

```python
class Foo: pass

class Bar[T: Foo]: ... # Which Foo does this refer to?

class Foo: pass # Foo is redefined here
```

@JelleZijlstra, in your implementation of PEP 695, it looks like you always defer evaluation of the bound expression for a type parameter. If that's the case, maybe we should update the spec to indicate that type checkers can rely on this behavior and assume that they should always use deferred evaluation for the bound (and constraint and default) expressions.

**Comment by JelleZijlstra** [comment]
Yes, the runtime implementation always defers evaluation of the bound. I agree it's worth changing the spec to clarify this. String annotations should still be allowed, like in any other type expression, but they are not necessary for forward references.

**Comment by JelleZijlstra** [comment]
It looks like I didn't incorporate that section of PEP 695 into the typing spec, because it deals primarily with the runtime implementation. The Python language reference is explicit that PEP 695 TypeVar bounds are lazily evaluated (https://docs.python.org/3.12/reference/executionmodel.html#lazy-evaluation), so I don't think this needs to be duplicated in the typing spec.

**Comment by erictraut** [comment]
OK, thanks for the clarification. I'll update pyright's implementation accordingly. 

I think we should also update the type checker conformance tests to validate this behavior. I'll add that to my to-do list as well.

**Comment by erictraut** [comment]
@JelleZijlstra, I presume that default value expressions (introduced in PEP 696 to extend the 695 syntax) are also deferred. Just want to confirm. BTW, are you implementing PEP 696 in the CPython code base? Just curious.

**Comment by JelleZijlstra** [comment]
Yes and yes. https://github.com/python/cpython/pull/116129

**Comment by erictraut** [comment]
This is included in pyright 1.1.359, which I just published.
